### PR TITLE
Add email verification

### PR DESCRIPTION
**Pull request by maggiekoesno**
Require email verification when creating new user

### CODE REVIEW

File: auth.py [hunk 1]
```diff
@@ -11,6 +11,40 @@
 
 auth = Blueprint('auth', __name__)
 
+def send_verification_email(user):
+    if not user.verified:
+        token = user.get_token(86400)
+        msg = Message('PetAdoption: Email Verification', sender='petadoption.sps@gmail.com',
```
Comment (hongshaoyang) [inline auth.py:26]: there are a couple of undefined variables in this file including `Message`, `account`, `mail`, `update_user`, `RequestVerificationMail`. could you use pylint to check the imports and fix them?

File: forms.py [hunk 2]
```diff
@@ -45,3 +45,7 @@ def validate_email(form, email_field):
         user = get_user(email_field.data)
         if user:
             raise ValidationError(f'Email {email_field.data} has already been taken, please use another email')
+
+class RequestVerificationEmail(FlaskForm):
+    email = StringField('Email', validators=[Email(), DataRequired()])
+    send = SubmitFiled('Resend Email')
```
Comment (hongshaoyang) [inline forms.py]: `SubmitField`

File: users.py [hunk 3]
```diff
@@ -1,31 +1,48 @@
...
-        return User(email, source['profile']['name'], source['profile']['hashed_password'], source['profile']['contact_number'], source['listings'])
+        return User(email, source['profile']['name'], source['profile']['hashed_password'], source['profile']['contact_number'], source['profile']['verified'], source['listings'])
 
     def to_firestore(self):
         return {
             "profile": {
                 "hashed_password": self.password,
                 "name": self.name,
                 "contact_number": self.contact_number
+                "verified": self.verified
```
Comment (hongshaoyang) [inline users.py:28]: missing comma before, use pylint to check these errors

File: users.py [hunk 4]
```diff
@@ -1,31 +1,48 @@
...
+    @staticmethod
+    def verify_token(token):
+        timed_serializer = TimedSerializer(app.config['SECRET_KEY'])
+        try:
+            user_email = timed_serializer.loads(token)['user_email']
+        except:
+            return None
+        return user_email
+
+from main import app
```
Comment (hongshaoyang) [inline users.py]: shift this import to the front

**Review by hongshaoyang — COMMENTED** [review]
Enable pylint and fix errors in changed files

File: auth.py [hunk 5]
```diff
@@ -11,6 +11,40 @@
...
+    if user is None:
+        flash('Your link has expired or is invalid! Request a new valid link!')
+        return redirect(url_for('auth.request_verification_email'))
+    user.verified = True
+    update_user(user.email, user.to_firestore())
+    return redirect(url_for('auth.login'))
+
+@auth.route('/request_verification_email', methods = ['GET', 'POST'])
+def request_verification_email():
+    form = RequestVerificationEmail()
```
Comment (hongshaoyang) [inline auth.py:39]: create the form by passing request.form into the constructor: `RequestVerificationEmail(request.form)`

File: templates/verify.html [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      <br>
+      {% with messages = get_flashed_messages() %}
+      {% if messages %}
+          <div class="notification is-danger">
+              {{ messages[0] }}.
+          </div>
+      {% endif %}
+      {% endwith %}
+      {% from "_formhelpers.html" import render_field %}
+      <form action="/login" method="post">
```
Comment (hongshaoyang) [inline templates/verify.html:19]: `action="/request_verification_email`

Comment (hongshaoyang) [inline templates/verify.html]: `action="/request_verification_email"`

File: auth.py (hunk 5 above)
Comment (hongshaoyang) [inline auth.py]: create the form by passing request.form into the creator: `RequestVerificationEmail(request.form)`

https://flask.palletsprojects.com/en/1.1.x/patterns/wtforms/#in-the-view

File: users.py (hunk 4 above)
Comment (hongshaoyang) [reply users.py]: you can use the 'organize imports' command or 'sort imports' command on vscode

Comment (maggiekoesno) [reply users.py]: This can't be shifted to the front, it has to be at the bottom to avoid circular imports

**Review by hongshaoyang — DISMISSED** [review]

**Comment by maggiekoesno** [comment]
There was an issue with email configs that I didn't notice before. My bad. Fixed it and should work now!

**Review by hongshaoyang — APPROVED** [review]

**Pull request merged**